rfcs/0000-expressive-diagnostics.md

```diff
@@ -0,0 +1,356 @@
+        info(
+            constant_decl.clone(),
+            "This is the original declaration of the imported constant \"{name}\"."
+        ),
+        error(
+            name.span(),
+            "Shadowing via {} \"{name}\" happens here.", 
+            if variable_or_constant == "Variable" { "variable" } else { "new constant" }
+        ),
+        help("Unlike variables, constants cannot be shadowed by other constants or variables."),
+        help(
+            match (variable_or_constant.as_str(), constant_decl.clone() != Span::dummy()) {
+                ("Variable", false) => format!("Consider renaming either the variable \"{name}\" or the constant \"{name}\"."),
+                ("Constant", false) => "Consider renaming one of the constants.".to_string(),
+                (variable_or_constant, true) => format!(
+                    "Consider renaming the {} \"{name}\" or using {} for the imported constant.",
+                    variable_or_constant.to_lowercase(),
+                    if *is_alias { "a different alias" } else { "an alias" }
+                ),
+                _ => unreachable!("We can have only the listed combinations: variable/constant shadows a non imported/imported constant.")
+            }
+        )
+    )]
+    ConstantsCannotBeShadowed {
+        variable_or_constant: String,
+        name: Ident,
+        constant_span: Span,
+        constant_decl: Span,
+        is_alias: bool,
+    },
+    ...
+}
+```
+
+_Issue_ and _infos_ point to places in code, and are commonly named _labels_. The places in code they refer to are given by `Span`s.
+`issue` and `info` elements of `#[error]` take `Span` as their first argument, as seen above. If the passed `Span` parameter equals `Span::dummy()` the _label_ is considered _not to be in source code_.
+
+For _hints_, not being in source code means that they are not rendered at all. Diagnostics API will ignore them and clients (CLI, LSP) will not be aware of them.
+
+For the _issue_, clients will always display it as a part of diagnostic description. If it is in code, clients can choose to additionally display it in code.
+E.g., `forc` always shows the _issue_ text as a part of the diagnostic description, but if the _issue_ is in code and there is a _hint_ pointing to the same place in code, the _issue_ will not be rendered as a _label_ in code. This overloading of the _issue_ text by a _hint_ is visible on the above example and is enforced via Diagnostics API. It gives us the flexibility to phrase the diagnostic description and the hints in an independent way, while allowing backward compatibility with the current warnings and errors that have only a single message and span.
+
+Using `Span::dummy()` to denote non-existance of an element is in line with the current usage of `Span::dummy()`. It also allows declarative definition of all the _hints_, knowing that some of them might not be shown, depending on the concrete `Span` value passed to the enum variant.
```

Comment:
I know that the LSP team is currently wondering if we could guarantee all spans have proper paths, so encouraging use of dummy spans may not be a good idea. We may want to use Option here.
@JoshuaBatty what do you think?

Reply:
TL:DR: I am all for `Option` but that would IMO mean getting rid of `Span::dummy` in general, or have a cumbersome checks and code every time we `emit_err` or `emit_warn`. 

In my first commit I left the comment saying something along the lines _// Should we use Option here?_. Strictly speaking, it should be an `Option`.

But after looking in the existing usages, it is clear that `Span::dummy()` is used to communicate non-existance. If we decide to use `Option` here, we should consistently use it everywhere. Which opens the question of the role of `Span::dummy()`.

Also, there is a good deal of convenience in using `Span::dummy()` in diagnostics because the existence of `Span` is the regular case, overwhelmingly regular. With `Span::dummy()`s we don't need to write `Some(my_span)` every time we define warning/error in code, but just `my_span`. Same with unwrappings etc. In those rare cases we know that we don't have span `Span::dummy()` sticks out nicely, better then `None` ;-)

Last but not  least, `Span::dummy()`s get propagated through the code now. When they reach diagnostic, we can just pass them over. If we don't switch everywhere to `Option<Span>` that would mean always checking if we got `Span::dummy()` and turning it into `None`. E.g. `Span::join_all()` returns `Span;:dummy()` if there are no spans.

Reply:
The nice thing about Option<Span> is that we could be explicit about which Warnings & Error must have a real span. Without this, bugs can easily creep in that are only surfaced in LSP. If we don't have tests for that particular edge case, they can be glaringly obvious to the end user.

My understanding is that most Warnings and Errors _should_ have a span so it would only need to be optional in some rare cases, but I could be wrong about that. It might be worth doing an audit of all the existing diagnostic types to see how many should _not_ require a real span.

Reply:
Yeah agree with Sophie. Have an Option instead of a Dummy span would be a lot better for the language server. We were running into a bug a few weeks ago where no diagnostics at all where showing in the editor. Turned out this was due to one of the tokens was assigned a DummySpan in the compiler. This means that the Span range was 0,0,0,0. If an LSP client receives a vector of Diagnostics and just one of them has a range of 0,0,0,0 then the client throws the entire Vec away.